onnxruntime/core/providers/coreml/builders/impl/cast_op_builder.cc

```diff
@@ -18,14 +19,49 @@ class CastOpBuilder : public BaseOpBuilder {
 
   bool HasSupportedInputsImpl(const Node& node, const OpBuilderInputParams& input_params,
                               const logging::Logger& logger) const override;
+
+ public:
+  bool SupportsMLProgram() const override { return true; }
 };
 
-Status CastOpBuilder::AddToModelBuilderImpl(ModelBuilder& /* model_builder */,
-                                            const Node& /* node */,
-                                            const logging::Logger& /* logger */) const {
-  // This is a special handling case for ArgMax Op, where argmax is followed by a cast to int32 type.
-  // The ArgMax is fused with the Cast node and produces an int32 output.
-  // Cast node is not provided in CoreML model, so we're skipping adding the Cast node here.
+Status CastOpBuilder::AddToModelBuilderImpl([[maybe_unused]] ModelBuilder& model_builder,
+                                            [[maybe_unused]] const Node& node,
+                                            [[maybe_unused]] const logging::Logger& logger) const {
+// This is a special handling case for ArgMax Op, where argmax is followed by a cast to int32 type.
+// The ArgMax is fused with the Cast node and produces an int32 output.
+#if defined(COREML_ENABLE_MLPROGRAM)
+  if (model_builder.CreateMLProgram()) {
+    using namespace CoreML::Specification::MILSpec;
+    // https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.elementwise_unary.cast
+
+    std::unique_ptr<Operation> op = model_builder.CreateOperation(node, "cast");
+    AddOperationInput(*op, "x", node.InputDefs()[0]->Name());
+    NodeAttrHelper helper(node);
+    auto cast_to_type = helper.Get("to", ONNX_NAMESPACE::TensorProto::UNDEFINED);
+    std::string to_dtype = "";
+    if (cast_to_type == ONNX_NAMESPACE::TensorProto::INT32 || cast_to_type == ONNX_NAMESPACE::TensorProto::INT64) {
+      to_dtype = "int32";
+      // CoreML doesn't support int64 while onnx used int64 in index input/output
+      // we have two solutions for this: 1) if cast is the first/last node in graph partition,
+      // we will convert the input/output to int32 forcely
+      // 2) if cast is a middle node of graph-parition, we just treat int64 as int32.
```

Comment:
There's probably a bit more detail that can be added. ONNX also allows int64 for input (e.g. token ids).

```suggestion
      // CoreML doesn't support int64, while ONNX uses int64 for indices and as well as data values.
      // We convert the data inputs/outputs between int64 and int32 when calling onnxruntime::coreml::Model::Predict,
      // and when adding int64 initializers to the CoreML model.
      // CoreML operators can only produce int32 and not int64 values.
      // Due to that there should be no actual int64 values inside the CoreML model and we can infer any 
      // ONNX_NAMESPACE::TensorProto::INT64 values to be int32.
```